nanshe_workflow/data.py

```diff
@@ -526,3 +527,300 @@ def astype(self, dtype):
+                with dask.set_options(get=self._client.get):
+                    dask.distributed.wait(old_value)
+
+        value = dask.array.from_array(
+            self._diskstore[key], chunks=old_value.chunks, name=name,
+            lock=self._disklock, fancy=False
+        )
+
+        with dask.distributed.client.temp_default_client(self._client):
+            with dask.set_options(get=self._client.get):
+                value = self._client.persist(value)
+                dask.distributed.wait(value)
+
+        del self._diskstore[key]
+
+        return value
+
+    def popitem(self):
+        """
+            Pops the next key-value pair by iter
+
+            Returns:
+                Dask Array: Key-Value pair that was in store
+
+            Note:
+                See ``pop`` for more details.
+        """
+        try:
+            key = next(iter(self))
+        except StopIteration:
+            raise KeyError("Store empty")
+        else:
+            value = self.pop(key)
+            return key, value
+
+    def clear(self):
+        """
+            Deletes all values in store
+
+            Note:
+                See ``__delitem__`` for more details.
+        """
+        while True:
+            try:
+                key = next(iter(self))
+            except StopIteration:
+                break
+            else:
+                del self[key]
+
+    def update(self, dct):
+        """
+            Updates keys with values in dictionary
+
+            Stores all non-redundant values to the disk store. Removes any
+            existing values on the way. Optimizes over the shared Dask graph
+            of all of the values provided reusing intermediate results as part
+            of the computation to store. Caches Dask Arrays pointing back to
+            the stored result for immediate reuse. Can track progress using
+            these Dask Arrays and/or chain them in other computations.
+
+            Args:
+                dct (Mapping): key-value pairs to store
+
+            Returns:
+                Dask Array: Value that was in store
+        """
+
+        keys = []
+        srcs = []
+        tgts = []
+        for k, v in dct.items():
+            v = dask.array.asarray(v)
+
+            # Skip if this Array is already stored
+            if self._memstore.get(k) is v:
+                continue
+
+            # Make sure the old key is removed before creating a new one
+            try:
+                del self[k]
+            except KeyError:
+                pass
+
+            chunks = tuple(max(c) for c in v.chunks)
+            v = v.rechunk(chunks)
+
+            # Salt keys to avoid referencing expired futures
+            # ref: https://github.com/dask/dask/issues/3322
+            v = v.map_blocks(
+                lambda a, u: a,
+                dtype=v.dtype,
+                token="salted",
+                u=uuid.uuid1().bytes
```

Comment:
We can improve this salting strategy if we construct a new graph with the salted keys, but will wait to see if it matters before trying that. Besides it would be better to solve this upstream.